server/src/test/java/org/elasticsearch/cluster/service/MasterServiceTests.java

```diff
@@ -982,6 +983,127 @@ public void onAckTimeout() {
+            masterService.submitStateUpdateTask("starved task", new ClusterStateUpdateTask(Priority.NORMAL) {
+                @Override
+                public ClusterState execute(ClusterState currentState) {
+                    assertFalse(keepRunning.get());
+                    starvedTaskExecuted.countDown();
+                    return currentState;
+                }
+
+                @Override
+                public void onFailure(String source, Exception e) {
+                    fail();
+                }
+            });
+
+            // check that a warning is logged after 5m
+            final MockLogAppender.EventuallySeenEventExpectation expectation1 = new MockLogAppender.EventuallySeenEventExpectation(
+                "starvation warning",
+                MasterService.class.getCanonicalName(),
+                Level.WARN,
+                "pending task queue has been nonempty for [5m/300000ms] which is longer than the warn threshold of [300000ms];" +
+                    " there are currently [2] pending tasks, the oldest of which has age [*"
+            );
+            mockAppender.addExpectation(expectation1);
+
+            while (relativeTimeInMillis - startTimeMillis <= warnThresholdMillis) {
+                awaitNextTask.run();
+                mockAppender.assertAllExpectationsMatched();
+            }
+
+            expectation1.setExpectSeen();
+            awaitNextTask.run();
+            // the master service thread is somewhere between completing the previous task and starting the next one, which is when the
+            // logging happens, so we must wait for another task to run too to ensure that the message was logged
+            awaitNextTask.run();
+            mockAppender.assertAllExpectationsMatched();
+
+            // check that another warning is logged after 10m
+            final MockLogAppender.EventuallySeenEventExpectation expectation2 = new MockLogAppender.EventuallySeenEventExpectation(
+                "starvation warning",
+                MasterService.class.getCanonicalName(),
+                Level.WARN,
+                "pending task queue has been nonempty for [10m/600000ms] which is longer than the warn threshold of [300000ms];" +
+                    " there are currently [2] pending tasks, the oldest of which has age [*"
+            );
+            mockAppender.addExpectation(expectation2);
+
+            while (relativeTimeInMillis - startTimeMillis <= warnThresholdMillis * 2) {
+                awaitNextTask.run();
+                mockAppender.assertAllExpectationsMatched();
+            }
+
+            expectation2.setExpectSeen();
+            awaitNextTask.run();
+            // the master service thread is somewhere between completing the previous task and starting the next one, which is when the
+            // logging happens, so we must wait for another task to run too to ensure that the message was logged
+            awaitNextTask.run();
```

Comment:
I am not sure I understand why the one wait one line up is not enough? The `awaitNextTask.run()` in the loop waits until the task that should trigger the warning has done its second await, the warning will be logged `afterExecute` and a single `awaitNextTask.run()` should be enough to ensure that it is logged?

Reply:
Ah, there's actually a subtle off-by-one error here, fixed by a22797334d2ce1d0c264a975eda89ac71a5a0f53. With that change, you should be able to observe a failure easily if you delete one of the `awaitNextTask.run()` lines. Without that change it also would fail, but only after ~1000 iterations when the message was logged early enough to fail the assertion.

The issue is that the test thread can only be certain that the logging happens after calling the first `awaitNextTask.run()` and before the second one returns. If we don't call `awaitNextTask.run()` twice in between setting the expectation up and checking it then we might miss it.

(Technically you don't have to set the expectation up before the message is logged, but you do have to skip the preceding `assertAllExpectationsMatched()` call)